### ISSUE TITLE
Inductor can return a tensor with different output strides, even with layout optimization disabled

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#116433]
This is the root cause of an issue with inductor + dynamic shapes (internal post [here](https://fb.workplace.com/groups/1075192433118967/permalink/7816560718370992/) with [repro](https://www.internalfb.com/intern/anp/view/?id=4767662)).

Normally, inductor tries to match the same output strides as eager, and only attempts to diverge when layout optimization is enabled (when it's determined to be good for perf).

However, it's possible for inductor's outputs to end up with strides different from eager, even if layout optimization is disabled. The divergence appears to be harmless, but it causes issues for dynamic shapes.

Here's a minimal example:
```
import torch

def f(x):
    tmp = x + 1
    return tmp.view(-1, 1, 2)

x = torch.arange(8, device='cuda', dtype=torch.float32)
out = f(x)
compiled_out = torch.compile(f)(x)
print(out.shape)  # []
print(compiled_out.shape)
print(out.stride())
print(compiled_out.stride())

# prints:
torch.Size([4, 1, 2])
torch.Size([4, 1, 2])
(2, 2, 1)
(2, 0, 1)
```

You can see that `output.stride()[1] == 2` in eager, but `0` with inductor. That's because it is harmless in this case: The corresponding dim has `output.size()[1] == 1`, so the stride at this dim is never used in any calculations to compute an offset into the tensor.

**Why is this a problem for dynamic shapes**

Today, layout optimization and dynamic shapes are known to not play well together, and so their combination is not allowed.

However, it looks like we:

(1) When we have forward outputs that are saved as activations for the backward graph, we detect when inductor gives us different strides from eager for those activations ([here](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py#L345))

(2) when this happens, we query inductor for the strides that it's using at runtime, and blat them directly into the backward graph [here](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py#L361)

(3) Once inductor sees strides that are static shapes, it is free to generate assertions that these shapes are fixed ([here](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/wrapper.py#L481))

The net result is that when inductor generates outputs in the forward with different strides from eager (even if harmless), we will end up with static strides in inductor, even if the user made it clear that their strides were dynamic. This results in the spurious stride assertion failure in the link above.

**Options for fixes?**

I can think of two:

(1) Force inductor not to use stride==0 when it sees that the corresponding dim has size==1

(2) Have our check for "did the strides diverge between inductor and eager" ignore this difference, since we know it is harmless and not the result of doing any layout optimizations.

cc @ezyang @msaroufim @wconstab @anijain2305 @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

**Comment by bdhirsh** [comment]
I'm tentatively going to push out a test patch for (2), since I think it will be easier to implement. Open to ideas though.

cc @eellison / @shunting314 since it's related to layout optimization

**Comment by bdhirsh** [comment]
also cc @ezyang for thoughts

**Comment by ezyang** [comment]
It would be nice if inductor respected the original user chosen stride in cases where it doesn't matter. It could materially make a difference for downstream eager mode layout propagation, where "don't care" strides actually can matter (see also https://github.com/pytorch/pytorch/issues/110387 ). I don't have a good sense for how easy it is to fix. Maybe it is easy and we just can do it.

**Comment by jansel** [comment]
This is the side effect of a performance optimization where the size=1/stride=0 dimensions will be optimized out of indexing expressions.

Should be easy to change the call to `empty_strided` to use the eager stride (while keeping the perf optimization in indexing).
